lib/models/statement.js

```diff
@@ -255,6 +273,198 @@ schema.statics.postFetchMap = function postFetchMap(statement) {
+    redisClient.get(`${prefix}-STARTED-AT`),
+    redisClient.get(`${prefix}-COMPLETED-AT`),
+    redisClient.get(`${prefix}-RESULTS`),
+  ]);
+
+  if (sinceAt && completedAt && moment(completedAt).isSameOrBefore(moment(sinceAt))) {
+    return {
+      result: null,
+      startedAt,
+      completedAt: null,
+    };
+  }
+
+  return {
+    result: results && JSON.parse(results),
+    startedAt,
+    completedAt,
+  };
+};
+
+/**
+ * @param {array} pipeline
+ * @param {object} authInfo
+ * @returns {array} decorated pipeline
+ */
+export const decoratePipeline = async (pipeline, authInfo) => {
+  const organisation = getOrgFromAuthInfo(authInfo);
+
+  const parsedPipeline = await parseQuery(pipeline, { organisation });
+
+  const scopedFilter = await getScopeFilter({
+    modelName: 'statement',
+    actionName: 'view',
+    authInfo,
+  });
+
+  parsedPipeline.unshift({
+    $match: scopedFilter
+  });
+
+  if (authInfo.client && authInfo.client.lrs_id) {
+    parsedPipeline.unshift({
+      $match: { lrs_id: objectId(authInfo.client.lrs_id) }
+    });
+  }
+
+  return removeEmptyMatch(parsedPipeline);
+};
+
+/**
+ * @param {*} authInfo
+ * @param {array} basePipeline
+ * @param {number} skip
+ * @param {number} limit
+ * @param {string|null} sinceAt
+ * @returns {Promise}
+ */
+export const aggregateAsync = async (
+  authInfo,
+  basePipeline,
+  skip,
+  limit,
+  sinceAt,
+) => {
+  const redisClient = redis.createClient();
+  const organisationId = getOrgFromAuthInfo(authInfo);
+  const pipeline = await decoratePipeline(basePipeline, authInfo);
+
+  const prefix = buildRedisKeyPrefix(organisationId, pipeline, skip, limit);
+
+  const { result, startedAt, completedAt } = await getAggregationStatus(redisClient, organisationId, pipeline, skip, limit, sinceAt);
+
+  // No cached
+  if (result === null) {
+    // if is running
+    if (startedAt) {
+      return { result, startedAt, completedAt };
+    }
+
+    // if is not running, start running query
+    await redisClient.set(`${prefix}-STARTED-AT`, moment().format('YYYY-MM-DD HH:mm:ss.SSSZ'));
+
+    // Do not await!
+    runAggregationAsync(redisClient, organisationId, pipeline, skip, limit);
+
+    await delay(ASYNC_AGGREGATION_WAITING_MS);
+
+    const result1 = await getAggregationStatus(redisClient, organisationId, pipeline, skip, limit, sinceAt);
+    return result1;
+  }
+
+  // if cached
+  const bestByDatetime = moment(completedAt).add(ASYNC_AGGREGATION_REFRESH_AT_SECONDS, 'seconds');
+  const isFresh = moment().isSameOrBefore(bestByDatetime);
+
+  if (isFresh) {
+    return { result, startedAt, completedAt };
+  }
+
+  // if is running
+  // if (moment(completedAt).isSameOrBefore(startedAt)) {
+  //   return {
+  //     result: null,
+  //     startedAt,
+  //     completedAt,
+  //   };
+  // }
+
+  // if cache is old, run aggregation query again
+  await redisClient.set(`${prefix}-STARTED-AT`, moment().format('YYYY-MM-DD HH:mm:ss.SSSZ'));
```

Comment:
This logic seems identical to the logic above, do you think it would be worth extracting this into a reusable function @h-kanazawa?

Perhaps there is a way to reorder this code so that we don't have to extract it. For example, we could make the first if block have the condition `result === null && startedAt`. Then have a second if block for `isFresh`. Finally we just run this code if it gets that far. I hope that's clear, I don't think I've explained that particularly well.